File: java-version/src/uno/models/Player.java
package uno.models;

import java.util.ArrayList;
import java.util.List;

/** A player (human or computer) and their hand of cards. */
public class Player {

    private final String name;
    private final List<Card> hand = new ArrayList<>();

    public Player(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public List<Card> getHand() {
        return hand;
    }

    /** Add a card to the hand (ignores null, e.g. an empty deck). */
    public void addCard(Card card) {
        if (card != null) {
            hand.add(card);
        }
    }

    /** Remove and return the card at index, or null if invalid. */
    public Card playCard(int index) {
        if (index >= 0 && index < hand.size()) {
            return hand.remove(index);
        }
        return null;
    }

    /** Number of cards in the hand. */
    public int handSize() {
        return hand.size();
    }

    /** Print the hand with indices so the human can choose. */
    public void showHand() {
        System.out.println(name + "'s hand:");
        for (int i = 0; i < hand.size(); i++) {
            System.out.println("  [" + i + "] " + hand.get(i));
        }
    }
}

File: python-version/src/uno/models/player.py
"""Game player (human or computer) and their hand of cards."""

from uno.models.card import Card


class Player:
    """Represents a player and manages their hand of cards."""

    def __init__(self, name: str):
        self.name = name
        self.hand: list[Card] = []

    def add_card(self, card: Card | None):
        """Add a card to the hand (ignores None, e.g. an empty deck)."""
        if card:
            self.hand.append(card)

    def play_card(self, index: int) -> Card | None:
        """Remove and return the card at `index`, or None if invalid."""
        if 0 <= index < len(self.hand):
            return self.hand.pop(index)
        return None

    def hand_size(self) -> int:
        """Number of cards in the hand."""
        return len(self.hand)

    def show_hand(self):
        """Print the hand with indices so the human can choose."""
        print(f"{self.name}'s hand:")
        for i, card in enumerate(self.hand):
            print(f"  [{i}] {card}")
